experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/schemaBuilder.ts

```diff
@@ -0,0 +1,127 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Assume, RestrictiveReadonlyRecord, transformObjectMap } from "../../../util";
+import { typeNameSymbol } from "../../contextuallyTyped";
+import { FieldKind } from "../../modular-schema";
+import { ImplicitFieldSchema, NormalizeField, SchemaBuilderBase } from "../../schemaBuilderBase";
+import {
+	FieldSchema,
+	InternalTypedSchemaTypes,
+	StructSchema,
+	TreeNodeSchema,
+	Unenforced,
+} from "../../typed-schema";
+import { ProxyNode, SharedTreeObject } from "./types";
+
+const factoryContent = Symbol("Node content");
+interface HasFactoryContent<T> {
+	[factoryContent]: T;
+}
+
+/**
+ * Returns the content stored on an object created by a {@link SharedTreeObjectFactory}.
+ */
+export function getFactoryContent(x: unknown): unknown | undefined {
+	return (x as Partial<HasFactoryContent<unknown>>)[factoryContent];
+}
+
+/**
+ * Creates `{@link SharedTreeObject}`s of some type via a `create` method.
+ * @alpha
+ */
+export interface SharedTreeObjectFactory<TSchema extends TreeNodeSchema<any, any>> {
+	/**
+	 * Create a {@link SharedTreeObject} that can be inserted into the tree via assignment `=`.
+	 * @param content - the data making up the {@link SharedTreeObject} to be created.
+	 * @remarks
+	 * The {@link SharedTreeObject} created by this function may _only_ be used for insertion into the tree.
+	 * It may not be read, mutated or queried in any way.
+	 */
+	create(
+		content: ProxyNode<Assume<TSchema, StructSchema>, "javaScript">,
+	): SharedTreeObject<Assume<TSchema, StructSchema>>;
+}
+
+class FactoryTreeNodeSchema<
+		Name extends string = string,
+		T extends
+			Unenforced<InternalTypedSchemaTypes.TreeSchemaSpecification> = InternalTypedSchemaTypes.TreeSchemaSpecification,
+	>
+	extends TreeNodeSchema<Name, T>
+	implements SharedTreeObjectFactory<TreeNodeSchema<Name, T>>
+{
+	public create(
+		content: ProxyNode<Assume<TreeNodeSchema<Name, T>, StructSchema>, "javaScript">,
+	): SharedTreeObject<Assume<TreeNodeSchema<Name, T>, StructSchema>> {
+		const node = {};
+		// Shallow copy the content and then add the type name symbol to it.
+		// The copy is necessary so that the input `content` object can be re-used as the contents of a different typed/named node in another `create` call.
+		const namedContent = { ...content, [typeNameSymbol]: this.name };
+		Object.defineProperty(node, factoryContent, { value: namedContent });
+		for (const [key] of this.structFields) {
```

Comment:
why don't we put the actual data in these properties? The API is much more type safe that way, and the code that reads in these objects to put them in the tree won't have to special case these factory objects.

Reply:
I like the sound of "type safe" and "no special case"! What exactly are you suggesting? The type returned here is a SharedTreeObject, and the type of the values of its properties are SharedTreeObject, SharedTreeMap, or SharedTreeList. If it were all objects, I think we could just move the data in and it would work okay, but what about maps and lists? Our internal code would continue to work since it's not type safe right now. It would be odd if a user ever started to walk one of these "fake" SharedTreeObjects, though, and noticed that they are readable... until they read a list, and they actually get a (contextually typed) array, rather than the SharedTreeList that the type says. We could also make "fake" versions of SharedTreeLists and SharedTreeMaps which allow reading and not writing (and also none of the stuff that will be behind Daniel's `node` symbol - e.g. event registration, etc.). But at that point, this idea becomes a bit more involved and it doesn't have much value for demo. That's why I put it as a TODO.

Is that what you meant, or did you have a different idea in mind?

Reply:
After some discussion, we decided to leave this as is, but in the future, we will make simple cases (where we can make a nice little data object that is the correct type) work the other way so they are fully type safe and the factor and direct literal assignment can work the same.